File: 程序设计训练/week3proj/spider/spider.go
package main

import (
	"fmt"
	"github.com/PuerkitoBio/goquery"
	"golang.org/x/net/html/charset"
	"net/http"
	"os"
	"regexp"
	"strings"
	"sync"
	"time"
	"unicode/utf8"
)

var cnt = 0
var mutex = sync.Mutex{}
var que = Queue{}
var vis = map[string]string{}
var goodPattern, _ = regexp.Compile(`people\.com\.cn/n1`)
var badPattern, _ = regexp.Compile(`(sports)|(tv)|(cpc)`)

func getHref(doc *goquery.Document) {
	defer func() {
		recover()
	}()
	as := doc.Find("a")
	mutex.Lock()
	defer mutex.Unlock()
	for i, a := range as.Nodes {
		text := as.Eq(i).Text()
		if utf8.RuneCountInString(text) >= 10 {
			for _, h := range a.Attr {
				if h.Key == "href" && goodPattern.MatchString(h.Val) && !badPattern.MatchString(h.Val) {
					_, ok := vis[h.Val]
					text := as.Eq(i).Text()
					if !ok && !strings.Contains(text, "图") {
						vis[h.Val] = as.Eq(i).Text()
						que.Push(h.Val)
					}
				}
			}
		}
	}
}

func getContent(doc *goquery.Document, fileName string) {
	defer func() {
		err := recover()
		if err != nil {
			os.Remove(fileName)
		} else {
			cnt++
		}
	}()
	content := doc.Find("#rwb_zw")
	_ = content.Nodes[0]    // check size
	t := doc.Find(".box01") // find class
	_ = t.Nodes[0]          // check size
	file, _ := os.Create(fileName)
	defer file.Close()
	file.Write([]byte(strings.TrimSpace(t.Text())))
	file.Write([]byte("\n"))
	file.Write([]byte(strings.TrimSpace(content.Text())))
}

func getDoc(url string) *goquery.Document {
	defer func() {
		recover()
	}()
	req, _ := http.NewRequest("GET", url, nil)
	req.Header.Add("User-Agent", "Mozilla/5.0 (Windows NT 6.1; WOW64; rv:12.0) Gecko/20100101 Firefox/12.0")
	client := &http.Client{Timeout: time.Second}
	res, _ := client.Do(req)
	reader, _ := charset.NewReader(res.Body, "")
	doc, _ := goquery.NewDocumentFromReader(reader)
	return doc
}

func Bfs(root string, n int) {
	const CoroNum = 50
	que.Push(root)
	chs := [CoroNum]chan bool{}
	for i := range chs {
		chs[i] = make(chan bool)
	}
	for ; !que.Empty() && cnt < n; {
		currentCoro := CoroNum
		if currentCoro >= que.Len() {
			currentCoro = que.Len()
		}
		for i := 0; i < currentCoro; i++ {
			go func(finish chan bool) {
				mutex.Lock()
				url := que.Pop()
				fmt.Println(url, cnt)
				mutex.Unlock()
				doc := getDoc(url)
				getHref(doc)
				mutex.Lock()
				fileName := vis[url]
				mutex.Unlock()
				getContent(doc, fileName)
				finish <- true
			}(chs[i])
		}
		for i := 0; i < currentCoro; i++ {
			<-chs[i]
		}
	}
}

func main() {
	os.Mkdir("data", os.ModePerm)
	os.Chdir("data")
	beg := time.Now()
	Bfs("http://www.people.com.cn/", 20000)
	fmt.Println("Get 10000 pages in", time.Now().Sub(beg))
}

const BufSize = 1 << 20

type Queue struct {
	head, tail int
	data       [BufSize]string
}

func (q *Queue) Push(value string) {
	q.data[q.tail] = value
	q.tail = (q.tail + 1) & (BufSize - 1)
}

func (q *Queue) Pop() string {
	ret := q.data[q.head]
	q.head = (q.head + 1) & (BufSize - 1)
	return ret
}

func (q *Queue) Empty() bool {
	return q.head == q.tail
}

func (q *Queue) Len() int {
	return (q.tail - q.head) & (BufSize - 1)
}

File: 程序设计训练/week3proj/spider/spider.py
import asyncio
import requests
import os
import bs4
from bs4 import BeautifulSoup
from urllib import request
from collections import deque
from typing import List, Tuple, NoReturn, Deque, Dict
import re

html_pattern = re.compile(r'<[^>]+>', re.S)
target_folder = 'data2'

good_pattern = re.compile(r'people\.com\.cn/n1')
bad_pattern = re.compile('(sports)|(tv)|(cpc)')


def get_content(soup: BeautifulSoup, file_name: str) -> NoReturn:
    try:
        text: bs4.element.Tag = soup.find_all(id='rwb_zw')[0]
        date: bs4.element.Tag = soup.find_all(class_='box01')[0]
    except Exception as e:
        raise e  # rethrow
    with open(os.path.join(target_folder, file_name), 'w', encoding='UTF-8') as f:
        f.write(date.text.strip())
        f.write('\n')
        for content in text.contents:
            para = html_pattern.sub('', str(content)).strip()
            if para and para.find('showPlayer') == -1:
                f.write(para)
                f.write('\n')


def get_href(soup: BeautifulSoup) -> Tuple[List[str], List[str]]:
    href_list: List[str] = []
    name_list: List[str] = []
    href: bs4.element.Tag
    for href in soup.find_all(name='a'):
        if len(href.contents) == 0 or len(href.contents[0]) < 12:
            continue
        if 'href' in href.attrs and href['href'].find('http') != -1 \
                and good_pattern.search(href['href']) and not bad_pattern.search(href['href']):
            href_list.append(href['href'])
            name_list.append(href.contents[0])
    return href_list, name_list


def bfs(root: str, n):
    que = deque([root], maxlen=n * 200)
    vis = {root: 'ROOT'}
    loop = asyncio.get_event_loop()

    async def work(url: str):
        try:
            nonlocal n
            print(url, n)
            urlopen = lambda u: request.urlopen(u, timeout=1)
            response = await loop.run_in_executor(None, urlopen, url)
            soup = BeautifulSoup(response.read(), 'lxml')
            href_list, name_list = get_href(soup)
            # print(href_list)
            for href, name in zip(href_list, name_list):
                if href not in vis:
                    vis[href] = name
                    que.append(href)
                    # print('appending...')
            get_content(soup, vis[url])
            n -= 1
        except Exception as e:
            pass
            # print(url, e)

    loop = asyncio.get_event_loop()
    while n > 0 and que:
        print('len of que', len(que))
        task_num = min(len(que), 50)
        tasks = [work(que.popleft()) for _ in range(task_num)]
        loop.run_until_complete(asyncio.wait(tasks))
    loop.run_until_complete(asyncio.sleep(0))
    loop.close()


def mkdir(path: str) -> NoReturn:
    path = path.strip().rstrip("\\")
    if not os.path.exists(path):
        os.makedirs(path)
    else:
        print('already exist')


mkdir(target_folder)
bfs('http://www.people.com.cn/', 20000)
